### DISCUSSION TITLE
Extending Tau?

### DISCUSSION THREAD

**Discussion by prologic** [NicoNex/tau#15] [General]
How about we add support for modules?

Maybe an FFI of some kind?

Looking at the design and implementation, it looks really clean, but slightly limited...

**Comment by NicoNex** [comment]
Hi @prologic, tanks for the suggestion, I appreciate your interest in Tau (btw I'm a big fan of bitcask).
Support for modules is currently work in progress and you can check it out in the [import](https://github.com/NicoNex/tau/tree/import) branch.
Currently there are 3 ways I'm experimenting with modules:

The first is the `import` builtin function which allows you to import external Tau code with the following syntax:
```python
myModule = import("myModule.tau")
```
This implementation works completely for the Eval implementation but is currently buggy with the VM.

The second one is the closest to the FFI you suggested and allows you to import Go plugins and work with them seamlessy:
```python
myPlugin = plugin("myPlugin.so")
```
With this implementation you can currently call functions, methods and access variables in the plugin, but for the moment it won't work with converting Tau types to channels, maps, structs, interfaces, arrays, and pointers, but the support for at least maps and arrays will be added soon. (For reference check out the [toValue function](https://github.com/NicoNex/tau/blob/import/obj/nativelibs.go#L144)).

The third is an initial experiment which I'll drop and consists of having a map with all the methods of a library built directly into the interpreter and it works using the `native` builtin function like this:
```python
regexp = native("regexp")
```
But as I said this implementation is not dynamic and only works with libraries compiled directly alongside the interpreter so this implementation will be dismissed.

Since this is my first experience at Programming Language Design the implementation of the import mechanism is going pretty slow so any kind of help is very appreciated even just in form of technical tips or resources I could study from (currently that's my own implementation) or feel free to send me a pull request.

**Comment by prologic** [comment]
I think the 1st and 2nd options are what we want here? I like the idea of writing Go plugins hat you can import with `my_plugin = plugin("/path/to/plugin")`.

Hopefully importing Tau code with `import()` is a bit more straightforward as you don't have to convert between types 🗡️

**Comment by prologic** [comment]
I managed to get [module support in monkey-lang working](https://git.mills.io/prologic/monkey-lang#modules) -- but I haven't touched that project in a really long time -- it was mostly an exercise in reading the aforementioned books and building the project in such a way others could learn from.

**Comment by prologic** [comment]
Btw I really like the design of Tau, I'm finding it quite elegant 👌

**Comment by NicoNex** [comment]
> I managed to get [module support in monkey-lang working](https://git.mills.io/prologic/monkey-lang#modules) -- but I haven't touched that project in a really long time -- it was mostly an exercise in reading the aforementioned books and building the project in such a way others could learn from.

Thank you, this is exactly what I was looking for since all other implementations work with import statements and not functions, so I'll definitely look into this and hopefully finally have a working implementation for the import.

**Comment by NicoNex** [comment]
> I think the 1st and 2nd options are what we want here? I like the idea of writing Go plugins hat you can import with `my_plugin = plugin("/path/to/plugin")`.
> 
> Hopefully importing Tau code with `import()` is a bit more straightforward as you don't have to convert between types 🗡️ 
> 

Luckily the the implementation of the plugins is working already so all I need to do is support the remaining go types that are supportable (I'm not sure whether I can support interfaces structs and pointers but I'll see what I can do).
So expect it to be merged into master soon.

**Comment by prologic** [comment]
Does supporting interfaces and structs make sense for FFI / Plugins?

**Comment by NicoNex** [comment]
Well sometimes go functions accept those as argument so it would be pretty cool to have a Tau to Go conversion.
Btw I forgot to mention that structs as return types from go functions are supported already but they act as a black box. 
This means that you don't have access to anything inside, not even exported fields but you can call their methods.

**Comment by prologic** [comment]
Does this match the semantics of [Objects](https://github.com/NicoNex/tau#objects) you've deigned?

**Comment by NicoNex** [comment]
> Does this match the semantics of [Objects](https://github.com/NicoNex/tau#objects) you've deigned?

Sorry, what do you mean?

**Comment by prologic** [comment]
Shouldn't struct fields be accessible?

**Comment by NicoNex** [comment]
Oh yeah of course, I think there's no way around that, so the solution will be limited anyway, although it might be useful especially for the "options" objects since they're pretty common and in those usually the fields are exported, but anyway this isn't really a priority.

Anyway yesterday I  moved the plugin code to a separate branch ([plugin](https://github.com/NicoNex/tau/tree/plugin)) and i think that once I've implemented the support for maps it will be ready to be merged already.

**Comment by NicoNex** [comment]
Hi @prologic, in the end the plugin system is ready and I've merged it on the master branch, I've also updated the README, if you want to have a look.
Any feedback would be very appreciated!

**Reply by prologic** [reply]
This is very nice! I'm quite excited about this! Now we just need to get `import()` in and I might have a play with making `tau` available in the userland of a pretty niche/custom Linux-based OS I'm working on ;) Do you prefer it be called `tau` or `t` as far as the binary goes and what you'd type into a shell to invoke Tau?

**Reply by NicoNex** [reply]
That's very nice, I think `tau` is good enough, after all it's short already and same goes for invoking it.
I currently have Tau installed via `go install` and to invoke it I simply type `tau -vm`.
Maybe should I make the VM implementation the default one? I have been considering this option for a while now.

**Reply by prologic** [reply]
Actually yes, can we make that the default? I'm also thinking of shebangs here where we _might_ have `#!/bin/tau` at the top of the file.

**Reply by prologic** [reply]
FWIW I did this too with monkey-lang whereby you use the eval-based interpreter to debug and flesh our new designs before then designing the bytecode and extending the "machine" -- But at the end of the day it becomes more of a "debug" or "dev" thing, the VM is always faster for end-users 😅

**Reply by NicoNex** [reply]
I agree, I find the Eval also useful to use Tau as library and as an embedded language since it makes the whole thing easier.
Some time ago I was experimenting with this feature and I made a Telegram bot that you can reprogram via Telegram directly (@tautotron_bot) and it makes use of the Eval part.
Although the Tau version it uses is a bit old now since that project is more of a proof of concept than anything.

**Comment by NicoNex** [comment]
Hey @prologic, the VM is now by default 👍🏻.

**Reply by prologic** [reply]
Oh nice! Sorry I haven't integrated tau yet (will let you know!) -- I've been side tracked with writing an `mkfs.ext` implementation (basically having to learn how the Ext2 filesystem is laid out on disk :D)

**Reply by NicoNex** [reply]
Oh nice, good luck with that! 😆

**Comment by NicoNex** [comment]
Hi @prologic, in the meanwhile that the import system is under development I added the [error handling](https://github.com/NicoNex/tau#error-handling) in the README.
It would be nice if you could give me some feedback on it 😄.

**Reply by prologic** [reply]
I can do that 👌

**Reply by prologic** [reply]
So... I read this last night in bed and and I have a few questions:

1. Why didn't you go with a tuple-style error handling like Go?
2. I'm assuming a function can either return it's result (whatever type that is) or an `error(...)`?

Am I understanding the semantics right?

**Reply by NicoNex** [reply]
Yup you're understanding the semantic.

I didn't go with go tuple style simply because I took advantage of Tau being weakly typed and this also allowed for a more functional error handling which is shorter to write.

So the function instead of having to return the zero value of the type and the error can simply return only the error.

After all if an error occurs the zero value is pretty useless. :)

**Reply by prologic** [reply]
Yeah I buy your argument 😅 Sounds and looks okay to me 👌

**Reply by prologic** [reply]
My only grip at the moment is this whole `isError()` which looks like a special builtin function? The reason I feel a bit "off" by this is I'm not aware of any other kind of Tau builtin like this? For example a general purpose `isType(value, type)` or something? Am I making sense?

**Reply by NicoNex** [reply]
How come it looks special?

Is it because of the name of the builtin or because in the example it has an assignment between the parenthesis?
I could change the name but I can't think of a better one, although I'm open to suggestions.

While about the assignment thing, I'm just taking advantage of the fact that every expression in Tau returns a value and so does the assignment.
The builtin `isError` just accepts an object and returns true if the said object is an error.
You could write it like this:
```python
res = div(16, 0)
if isError(res) {
    exit(res)
}
println(res)
```
or shorter
```python
if isError(res = div(16, 0)) {
    exit(res)
}
println(res)
```

The latter is more similar to a try-catch block where in the try you have an expression and in the catch you handle the exception.

There is a builtin `type` that returns the type of the object that you pass to it as string, so the same thing can be achieved with:
```python
res = div(16, 0)
if type(res) == "error" {
    exit(res)
}
println(res)
```
or

```python
if type(res = div(16, 0)) == "error" {
    exit(res)
}
println(res)
```
But i feel that's more cumbersome.

**Reply by prologic** [reply]
Hmmm lemme have a re-read again. This is always tough ;)

**Reply by prologic** [reply]
Okay I _think_ the issue I have is `isError` seems to be the only builtin you've defined recently  "camelCase" style. Every other builtin is "lowercase".

**Reply by prologic** [reply]
Just playing around with an idea, but it doesn't work 😅

```#!diff
diff --git a/obj/builtin.go b/obj/builtin.go
index 9f4253f..ae18b54 100644
--- a/obj/builtin.go
+++ b/obj/builtin.go
@@ -106,6 +106,15 @@ var Builtins = []struct {
 			return NewString(args[0].Type().String())
 		},
 	},
+	{
+		"is",
+		func(args ...Object) Object {
+			if l := len(args); l != 2 {
+				return NewError("is: wrong number of arguments, expected 2, got %d", l)
+			}
+			return NewBoolean(args[0].Type() == args[1].Type())
+		},
+	},
 	{
 		"int",
 		func(args ...Object) Object {
@@ -346,17 +355,6 @@ var Builtins = []struct {
 			return NewClass()
 		},
 	},
-	{
-		"isError",
-		func(args ...Object) Object {
-			if l := len(args); l != 1 {
-				return NewError("isError: wrong number of arguments, expected 1, got %d", l)
-			}
-
-			_, ok := args[0].(*Error)
-			return ParseBool(ok)
-		},
-	},
 	{
 		"plugin",
 		func(args ...Object) Object {

```

For example I was thinking something like:

```#!tau
>>> x = 4
>>> is(x, int)
true
>>> foo = error("foo")
>>> is(foo, error)
true
```

But of course this doesn't work, finally getting into Tau's codebase and understanding the implementation/design 🙃

**Reply by prologic** [reply]
Really lov'n Tau btw 😅

**Reply by prologic** [reply]
Maybe rename `isError` to just `failed`?

e.g:

```#!diff
diff --git a/obj/builtin.go b/obj/builtin.go
index 9f4253f..31f5b6f 100644
--- a/obj/builtin.go
+++ b/obj/builtin.go
@@ -347,10 +347,10 @@ var Builtins = []struct {
 		},
 	},
 	{
-		"isError",
+		"failed",
 		func(args ...Object) Object {
 			if l := len(args); l != 1 {
-				return NewError("isError: wrong number of arguments, expected 1, got %d", l)
+				return NewError("failed: wrong number of arguments, expected 1, got %d", l)
 			}
 
 			_, ok := args[0].(*Error)

```

**Reply by NicoNex** [reply]
Happy to hear you're looking into the codebase, since even though Tau is inspired by Monkey, is still very different in the implementation.

The `failed` built-in isn't bad, I might actually change it since I wasn't completely ok with the camel case either. 😄

**Reply by prologic** [reply]
Sounds good to me! 👌 It also has a nice "ring" to it too when you read the code. Something I always aspire for is, code is read more often than it is written. Something I believe in 😂

**Comment by NicoNex** [comment]
Hey @prologic, now the plugin system also supports Tau to Go struct and pointer conversion.
It has trouble with custom types that have an underlying type though.

**Reply by prologic** [reply]
Nice! Did you write and include an example plugin btw? I haven't plyaed with the plugin system yet...

**Reply by NicoNex** [reply]
I didn't include any example plugin, personally I'm testing with [echotron](https://github.com/NicoNex/echotron) which is a library I've written since it has almost all use cases. 😄

**Reply by NicoNex** [reply]
Btw do you by any chance have Telegram? Or you only use Salty IM? 😆

**Reply by prologic** [reply]
I only use Signal (for now), Salty IM of course and IRC 😆

**Reply by prologic** [reply]
> I didn't include any example plugin, personally I'm testing with [echotron](https://github.com/NicoNex/echotron) which is a library I've written since it has almost all use cases. 😄 

Can you chick in a "Hello World" plug-in showing off the basics?

**Reply by NicoNex** [reply]
I can but the problem with plugins is that you need them to be compiled with the same compiler version as Tau, so I'm wondering if I should include a pre-compiled one or simply include some go code and show how it works.
For the moment though there's a small example in the README but it doesn't show all that you can do.

**Reply by prologic** [reply]
I'll do it and put up a PR in a few mins 👌
